cmd/sgbouncewizard/main.go

```diff
@@ -55,12 +136,125 @@ func ListRules(w http.ResponseWriter, r *http.Request) {
 
 func generateHash(pwd []byte) string {
 	hash, err := bcrypt.GenerateFromPassword(pwd, bcrypt.MinCost)
+	if err != nil {
+		log.Println(err)
+		return "error"
+	}
+	return string(hash)
+}
+
+func createRuleDB(rule *BounceRule) []byte {
+	db, err := sql.Open("mysql", "root:root@tcp(127.0.0.1:3306)/drop_rules")
+	checkErr(err)
+
+	stmt, err := db.Prepare("INSERT INTO bounce_rule(id,response_code,enhanced_code,regex,priority,description,bounce_action) VALUES(?,?,?,?,?,?,?)")
+	checkErr(err)
+
+	_, err = stmt.Query(rule.Id, rule.Response_code, rule.Enhanced_code, rule.Regex, rule.Priority, rule.Description, rule.Bounce_action)
+	checkErr(err)
+
+	defer stmt.Close()
+	defer db.Close()
+
+	data, err := json.Marshal(rule)
+	checkErr(err)
+	return data
+}
+
+func getRuleDB(id string) (*BounceRule, error) {
+	db, err := sql.Open("mysql", "root:root@tcp(127.0.0.1:3306)/drop_rules")
+	var bounce_rule *BounceRule
+
+	checkErr(err)
+	rows, err := db.Query("SELECT * FROM bounce_rule WHERE id=" + id)
+
+	for rows.Next() {
+		var br BounceRule
+		err = rows.Scan(&br.Id, &br.Response_code, &br.Enhanced_code, &br.Regex, &br.Priority, &br.Description, &br.Bounce_action)
+		checkErr(err)
+		bounce_rule = &br
+		return bounce_rule, nil
+	}
+	checkErr(err)
+
+	defer rows.Close()
+	defer db.Close()
+	return nil, err
+}
 
+func deleteRuleDB(id int) (*BounceRule, error) {
+	db, err := sql.Open("mysql", "root:root@tcp(127.0.0.1:3306)/drop_rules")
+	checkErr(err)
+	query := fmt.Sprintf("%s%d", "DELETE FROM bounce_rule WHERE id=", id)
+	_, err = db.Query(query)
+	checkErr(err)
+	defer db.Close()
+	return nil, err
+}
+
+func updateRuleDB(ruleDifferences map[string]interface{}, prevRule *BounceRule) {
+	queryString := createUpdateQuery(ruleDifferences, prevRule)
+	db, err := sql.Open("mysql", "root:root@tcp(127.0.0.1:3306)/drop_rules")
+	checkErr(err)
+	_, err = db.Query(queryString)
+	checkErr(err)
+	db.Close()
+}
+
+// Helper functions
+
+func checkErr(err error) {
 	if err != nil {
 		log.Println(err)
 	}
+}
 
-	return string(hash)
+func getRuleDifferences(prevRule *BounceRule, newRule *BounceRule) map[string]interface{} {
+	ruleChange := make(map[string]interface{})
+
+	prevRuleValue := reflect.ValueOf(prevRule).Elem()
```

Comment:
Why are we using reflection here? Don't we know the element types of `BounceRule`?